frontend/dockerfile/builder/build.go

```diff
@@ -196,12 +194,27 @@ func filter(opt map[string]string, key string) map[string]string {
 	return m
 }
 
-func parseGitSource(ref string) llb.State {
-	ref = strings.TrimPrefix(ref, gitPrefix)
+func detectGitContext(ref string) (*llb.State, bool) {
+	found := false
+	if httpPrefix.MatchString(ref) && gitUrlPathWithFragmentSuffix.MatchString(ref) {
+		found = true
+	}
+
+	for _, prefix := range []string{"git://", "github.com/", "git@"} {
```

Comment:
If it weren't for the `github.com/` (which I think is/should be `Dockerfile` specific) in here this would be quite a convenient utility function for the common case, is there some way we could extract/abstract it perhaps?

A new `SourceFromPath` function which produced either an `llb.Git` or an `llb.Local` depending what it found might be useful?